doc/shared/sundials/Errors.rst

```diff
@@ -0,0 +1,186 @@
+.. ----------------------------------------------------------------
+   SUNDIALS Copyright Start
+   Copyright (c) 2002-2023, Lawrence Livermore National Security
+   and Southern Methodist University.
+   All rights reserved.
+
+   See the top-level LICENSE and NOTICE files for details.
+
+   SPDX-License-Identifier: BSD-3-Clause
+   SUNDIALS Copyright End
+   ----------------------------------------------------------------
+
+.. _SUNDIALS.Errors:
+
+Error Checking
+==============
+
+.. versionadded:: 7.0.0
+
+Errors that occur within SUNDIALS are dealt with through error codes and error handler callback
+functions. The error code type :c:type:`SUNErrCode` is just a typedef to an ``int``:
+
+.. c:type:: typedef int SUNErrCode 
+
+Errors are always negative, while success is defined with the ``SUN_SUCCESS`` code and has the value ``0``.
+To see all of the possible error codes, refer to the ``sundials/sundials_errors.h`` header file 
+(`here <https://github.com/LLNL/sundials/blob/main/include/sundials/sundials_errors.h>_`).
+
+Functions in the SUNDIALS core API (i.e., ``SUN`` or ``N_V`` functions only) either return a
+:c:type:`SUNErrCode`, or (if they don't return a :c:type:`SUNErrCode`) they internally record an
+error code (if an error occurs) within the :c:type:`SUNContext` for the execution stream. 
+This "last error" is accessible via the :c:func:`SUNContext_GetLastError` or
+ :c:func:`SUNContext_PeekLastError` functions.
+
+Thus, in user code, SUNDIALS core functions can be checked for errors in one of two ways:
+
+.. code-block:: C
+
+  SUNContext sunctx;
+  SUNErrCode sunerr; 
+  N_Vector v;
+  int length;
+  sunrealtype dotprod;
+  
+  // Every code that uses SUNDIALS must create a SUNContext.
+  sunctx = SUNContext_Create(...);
+
+  // Create a SUNDIALS serial vector.
+  // Some functions do not return an error code.
+  // We have to check for errors in these functions using SUNContext_GetLastError.
```

Comment:
Checking the error is better but we should note there are cases where a context is not available and we need to check for a `NULL` return. We should document these exceptions somewhere.

Reply:
I think there are only a handful of functions that our in the user space that they might call that don't have a SUNContext and dont return an error code. I think we should probably change the signature of these functions to return an error code so that we dont have to tell users that there are exceptions to the error checks.

Reply:
I agree that we should document the exceptions, but my point is, most are internal so we should document the exceptions in the developer docs and eliminate the handful of exceptions in the user spac.e

Reply:
Agreed, it's best to keep the usage consistent for users and keep the exceptions internal if possible.

Reply:
I guess there are more than I thought. E.g., Nothing in `sundials_band.h` has a SUNContext and the calls do not return an error code. 

I just really dislike the idea of asking users to be aware of when to check NULL vs when to check an error code.  It would be great, if we could think of an acceptable way to avoid this.

Reply:
I think files like `sundials_band.h` and `sundials_dense.h` are things we should consider making internal at some point. Hopefully, that would help make the number of exceptions more manageable.

Reply:
Should we deprecate or remove those now?

Reply:
It doesn't need to be in this PR but I think we should. That said, let's discuss tomorrow.

Reply:
We'll need to update our examples that directly `#include` those to use in preconditioning (I believe that CVODE and CVODES still include `sundials_dense.h`.  I think that I already ported a corresponding ARKODE example to use the SUNLinearSolver_Dense module instead, so perhaps that could be used as a template.

Reply:
This is everything I could find in our public API that does not have access to a `SUNContext` and does not return a `SUNErrCode`.

```c
N_Vector* N_VNewVectorArray(int count);

SUNMemory SUNMemoryHelper_Alias(SUNMemory mem);
SUNMemory SUNMemoryHelper_Wrap(void* ptr, SUNMemoryType mem_type);

// The functions in sundials_math.h
sunrealtype SUNRpowerI(sunrealtype base, int exponent);
sunrealtype SUNRpowerR(sunrealtype base, sunrealtype exponent);
sunbooleantype SUNRCompare(sunrealtype a, sunrealtype b);
sunbooleantype SUNRCompareTol(sunrealtype a, sunrealtype b, sunrealtype tol);
sunrealtype SUNStrToReal(const char* str);

int SUNQRsol(int n, sunrealtype **h, sunrealtype *q, sunrealtype *b);
int SUNQRfact(int n, sunrealtype **h, sunrealtype *q, int job);

FILE* SUNDIALSFileOpen(const char* filename, const char* modes);
void SUNDIALSFileClose(FILE* fp);

// Everything in sundials_direct.h
// Everything in sundials_dense.h
// Everything in sundials_band.h
```

We could create updated versions of all or some of these functions that take a SUNContext as an argument (except those in `sundials_direct.h`, `sundials_dense.h` which would be better deprecated it sounds like).

Reply:
Ill note that I left constructors out of these because they do have access to a `SUNContext`, it just might be `NULL`, which is a different scenario. We could handle this scenario with a static (global) error handler or consider it as catastrophic error that we don't handle.

Reply:
I think making `sundials_math.h` would be good to do as well. When needed, we can define equivalent macros in the examples (we have some that do this already to avoid using functions from `sundials_math.h`).

If `tgmath.h` is available, such macros in the examples would be unnecessary but I think MSVC requires C11 for that.

Reply:
OK, I noted that

```
// The functions in sundials_math.h
sunrealtype SUNRpowerI(sunrealtype base, int exponent);
sunrealtype SUNRpowerR(sunrealtype base, sunrealtype exponent);
sunbooleantype SUNRCompare(sunrealtype a, sunrealtype b);
sunbooleantype SUNRCompareTol(sunrealtype a, sunrealtype b, sunrealtype tol);
sunrealtype SUNStrToReal(const char* str);
// Everything in sundials_direct.h
// Everything in sundials_dense.h
// Everything in sundials_band.h
```

will all be deprecated.  Then, since this is a major release and these functions are rarely used, I have updated

```c
N_Vector* N_VNewVectorArray(int count);

SUNMemory SUNMemoryHelper_Alias(SUNMemory mem);
SUNMemory SUNMemoryHelper_Wrap(void* ptr, SUNMemoryType mem_type);

FILE* SUNDIALSFileOpen(const char* filename, const char* modes);
void SUNDIALSFileClose(FILE* fp);
```

to either return a `SUNErrCode` or have access to a `SUNContext`. 

I will update functions from `sundials_iterative.h` in the follow-on PR as they require changing some examples that provide ATimesFn etc.